pkg/providers/v1/zones.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"k8s.io/klog/v2"
+)
+
+type zoneDetails struct {
+	name     string
+	id       string
+	zoneType string
+}
+
+type zoneCache struct {
+	cloud             *Cloud
+	mutex             sync.Mutex
+	zoneNameToDetails map[string]zoneDetails
+}
+
+// Get the zone details by zone names and load from the cache if available as
+// zone information should never change.
+func (z *zoneCache) getZoneDetailsByNames(zoneNames []string) (map[string]zoneDetails, error) {
+	if len(zoneNames) == 0 {
+		return map[string]zoneDetails{}, nil
+	}
+
+	z.mutex.Lock()
+	defer z.mutex.Unlock()
+
+	if z.shouldPopulateCache(zoneNames) {
+		// Populate the cache if it hasn't been populated yet
+		err := z.populate()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	requestedZoneDetails := map[string]zoneDetails{}
+	for _, zone := range zoneNames {
+		if zoneDetails, ok := z.zoneNameToDetails[zone]; ok {
+			requestedZoneDetails[zone] = zoneDetails
+		} else {
+			klog.Warningf("Could not find zone %s", zone)
+		}
+	}
+
+	return requestedZoneDetails, nil
+}
+
+// NOTE: This method is not thread safe and should not be called outside of getZoneDetailsByNames
```

Comment:
You could acquire the lock again instead of relying on the good behavior of the caller :)

You can also use a `RWMutex` if you want to get fancy, though I don't expect us to have much contention between readers and writers in this case.

Reply:
Ya, I'm of two minds.

1. The purpose of these two methods was to tidy up `getZoneDetailsByNames`, so I preferred the cleanliness of only acquiring the lock in the parent and keeping the children blissfully ignorant.
2. While it's likely that we may never have a need to call these again, it's not a guarantee, and someone could technically write some code that's not thread safe and nobody reads the warnings on these methods.

Anywho, this probably won't be a decision ever visited again, but the conservative thing to do would be to remove the potential foot gun...